sklearn/tree/_splitter.pyx

```diff
@@ -481,9 +481,12 @@ cdef class BestSplitter(BaseDenseSplitter):
 
                             if current_proxy_improvement > best_proxy_improvement:
                                 best_proxy_improvement = current_proxy_improvement
-                                current.threshold = (Xf[p - 1] + Xf[p]) / 2.0
+                                # sum of halves is used to avoid infinite value
+                                current.threshold = Xf[p - 1] / 2.0 + Xf[p] / 2.0
 
-                                if current.threshold == Xf[p]:
+                                if ((current.threshold == Xf[p]) or
+                                    (current.threshold == INFINITY) or
+                                    (current.threshold == -INFINITY)):
```

Comment:
Should perhaps one of the infinite cases be the other direction (Xf[p] instead of Xf[p-1])? Or does it matter too little?

Reply:
I think it should always be in the direction of Xf[p-1] because that is what the split give us, the first p-1 elements to the left, the rest to the right. If we take Xf[p] then we will have p elements on the left.